### PR TITLE
Decouple Kubernetes Version from the releases of KET

### PR DESCRIPTION
**Pull request by dkoshkin**
Implementation for https://github.com/apprenda/kismatic/pull/1038

### CODE REVIEW

**Comment by swade1987** [comment]
@dkoshkin would be great to see some CLI output examples of the following:

- Attempting to go from 1.8.x to 1.9.x
- Attempting to go from 1.9.0 to 1.9.x
- Attempting to go from 1.9.x to 1.8.x
- Attempting to go from 1.9.x to 1.9.0

**Comment by dkoshkin** [comment]
What do you mean `go from`, upgrades?

**Comment by emmetthitz** [comment]
@swade1987 I don't think it would be any different than a typical upgrade. Should just be a `./kismatic upgrade`

**Comment by swade1987** [comment]
@emmetthitz  agreed but just wanted to make sure. 

Also wanted to make sure we have a list of scenarios we are testing against.

Those are the ones I came up with off the top of my head.

File: pkg/inspector/cmd/client.go [hunk 1]
```diff
@@ -39,13 +42,22 @@ func NewCmdClient(out io.Writer) *cobra.Command {
 			}
 			// Set the target node as the first argument
 			opts.targetNode = args[0]
+			opts.additionalVariables = make(map[string]string)
```
Comment (emmetthitz) [inline pkg/inspector/cmd/client.go]: I'm not sure how I feel about this. What was the motivation behind it?

File: pkg/install/plan_types.go [hunk 2]
```diff
@@ -85,6 +85,10 @@ type Cluster struct {
 	// cluster name, such as kubeconfig files and certificates.
 	// +required
 	Name string
+	// The Kubernetes version to install.
+	// If left blank will be set to the latest tested version.
+	// +default=v1.9.1
+	Version string
```
Comment (emmetthitz) [inline pkg/install/plan_types.go]: Does semver not serialize the way you'd expect?

File: pkg/install/version.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	}
+	latest := strings.Trim(string(body), " \t\n")
+	if !kubeReleaseRegex.MatchString(latest) {
+		return kubernetesVersion, kubernetesVersionString, fmt.Errorf("Invalid version format %q", latest)
+	}
+	parsedLatest, err := parseVersion(latest)
+	if err != nil {
+		return kubernetesVersion, kubernetesVersionString, fmt.Errorf("Could not parse version string %q", latest)
+	}
+	return parsedLatest, latest, nil
```
Comment (emmetthitz) [inline pkg/install/version.go]: Related to previous comment, feels weird to return a semver.Version and a string representation of the same thing.

File: docs/plan-file-reference.md [hunk 4]
```diff
@@ -2,6 +2,7 @@
 ## Index
 * [cluster](#cluster)
   * [name](#clustername)
+  * [version](#clusterversion)
```
Comment (swade1987) [inline docs/plan-file-reference.md]: I think should be `kubernetes_version` as it feels more accurate as you aren't versioning the cluster itself.

Comment (dkoshkin) [reply docs/plan-file-reference.md]: This gets autogenerated and thats just the href link, its actually just `cluster.version` see [here](https://github.com/apprenda/kismatic/pull/1048/files/a7a7d8a7401419b1aa3982412837ec51065cad3e#diff-e93a0c402f6331148e311cfe5b9e94a0R91)

Comment (swade1987) [reply docs/plan-file-reference.md]: I still hold it should really be `kubernetes_version`.

**Comment by swade1987** [comment]
**1. Online Scenarios:** 

O1 - Performing `kismatic install plan` what is the default value set to?

O2 - Performing a `kismatic install apply` and setting a version (A.B.C) that doesn't exist yet.

O3 - Performing a `kismatic install apply` and setting a version of (A.B.C-alpha.D) 

O4 - Performing a `kismatic install upgrade` and setting a version (A.B.C) that doesn't exist yet.

O5 - Performing a `kismatic install upgrade` and setting a version of (A.B.C-alpha.D)

**2. Offline Scenarios:** 

OF1 - Performing `kismatic install plan` what is the default value set to?

OF2 - Performing a `kismatic install apply` and setting a version (A.B.C) that doesn't exist yet.

OF3 - Performing a `kismatic install apply` and setting a version of (A.B.C-alpha.D) 

OF4 - Performing a `kismatic install upgrade` and setting a version (A.B.C) that doesn't exist yet.

OF5 - Performing a `kismatic install upgrade` and setting a version of (A.B.C-alpha.D)


@willvrny not sure if you have any more scenarios to cover?

**Comment by dkoshkin** [comment]
1. Online Scenarios:

O1 - Performing kismatic install plan what is the default value set to?
Plan file gets set to tested version
```
cluster:
  name: kubernetes

  # Kubernetes cluster version
  version: v1.9.1
```
O2 - Performing a kismatic install apply and setting a version (A.B.C) that doesn't exist yet.
```
➜  out git:(decouple-k8s) ./kismatic install validate

Validating==========================================================================
Reading installation plan file "kismatic-cluster.yaml"                          [OK]
Validating installation plan file                                               [ERROR]
- Cluster version "v1.9.4" greater than the latest stable version "v1.9.2"
Plan file validation error prevents installation from proceeding
```
O3 - Performing a kismatic install apply and setting a version of (A.B.C-alpha.D)
Setting in plan file:
```
  # Kubernetes cluster version
  version: v1.9.3-beta.0
```
```
➜  out git:(decouple-k8s) ./kismatic install validate

Validating==========================================================================
Reading installation plan file "kismatic-cluster.yaml"                          [OK]
Validating installation plan file                                               [ERROR]
- Cluster version "v1.9.3-beta.0" must be a valid version, ie "v1.9.1"
Plan file validation error prevents installation from proceeding
```
O4 - Performing a kismatic install upgrade and setting a version (A.B.C) that doesn't exist yet.
```
➜  out git:(decouple-k8s) ./kismatic upgrade online

Computing upgrade plan==============================================================
Reading plan file                                                               [OK]
Validating installation plan file                                               [ERROR]
- Cluster version "v1.9.4" greater than the latest stable version "v1.9.2"
Plan file validation error prevents installation from proceeding
```
O5 - Performing a kismatic install upgrade and setting a version of (A.B.C-alpha.D)
```
➜  out git:(decouple-k8s) ./kismatic upgrade online

Computing upgrade plan==============================================================
Reading plan file                                                               [OK]
Validating installation plan file                                               [ERROR]
- Cluster version "v1.9.3-beta.0" must be a valid version, ie "v1.9.1"
Plan file validation error prevents installation from proceeding
```
2. Offline Scenarios:
When `disconnected_installation` is set to `true`, the `cluster.version` string is not verified with `https://storage.googleapis.com/kubernetes-release/release`.

However the inspector will validate that the packages exist in the repos and fail if not.

File: pkg/install/plan_types.go (hunk 2 above)
Comment (dkoshkin) [reply pkg/install/plan_types.go]: semver does not support having the `v` prefix, we want to be able to properly set the k8s version that includes the `v` prefix

File: pkg/inspector/cmd/client.go (hunk 1 above)
Comment (dkoshkin) [reply pkg/inspector/cmd/client.go]: We need to be able to pass in the version of k8s to test for(and other versions in the future). The inspector has different types of rules that a user can write, with this change the user is able to pass in arbitrary values and template them in the rules.

**Comment by swade1987** [comment]
@dkoshkin like it!

What about if I set the version to 1.10.1 (which exists) if I am running 1.9.1.

What is the error message then?

**Comment by dkoshkin** [comment]
In both cases when setting to 1.8.x or 1.10.x it would return 
```
➜  out git:(decouple-k8s) ./kismatic install validate

Validating==========================================================================
Reading installation plan file "kismatic-cluster.yaml"                          [OK]
Validating installation plan file                                               [ERROR]
- Cluster version "v1.10.1" must be a valid version, ie "v1.9.1"
Plan file validation error prevents installation from proceeding
```

**Comment by swade1987** [comment]
If I were a user I would find that error message misleading as it is valid.

It may be better to say something like ...

`Cluster version "v1.10.1" must be a valid version, ie "v1.9.x" this release doesn't support upgrading to a new minor release`

**Comment by swade1987** [comment]
How about `Cluster version "v1.10.1" invalid, must be a valid 1.9.x version, ie "v1.9.1"`

**Comment by swade1987** [comment]
How about the following command:

```
# Kubernetes cluster version (supported versions 1.9.x)
```

**Comment by dkoshkin** [comment]
Modified the error messages as discussed
```
➜  kismatic install validate

Validating==========================================================================
Reading installation plan file "kismatic-cluster.yaml"                          [OK]
Validating installation plan file                                               [ERROR]
- Cluster version "v1.9.5" invalid, the latest stable version is "v1.9.2"
Plan file validation error prevents installation from proceeding

➜  kismatic install validate

Validating==========================================================================
Reading installation plan file "kismatic-cluster.yaml"                          [OK]
Validating installation plan file                                               [ERROR]
- Cluster version "v1.8.5" invalid, must be a valid "v1.9.x" version, ie "v1.9.2"
Plan file validation error prevents installation from proceeding
```

**Comment by swade1987** [comment]
@dkoshkin LGTM!

**Pull request merged**